Add test to robot-name to make sure name in reset isn't hard-coded (fixes #148)

nwallace opened the pull request:
As per [this discussion](https://github.com/exercism/xclojure/issues/148), the tests didn't prevent you from hard-coding the new name in the `reset-name` function. This adds a new test to make sure you don't do that.

nwallace commented:
Hm, I'm not able to reproduce the error from Travis.  Can anybody else?

kytrinyx commented:
Huh. Not sure! @canweriotnow would you have a chance to check this? (Also, we're up against the end of Hacktoberfest. Could you review/merge before the end of October to give @nwallace credit for the PR in time to get a t-shirt?)

yurrriq commented:
I can take a look today. I think we should maybe try to avoid the top-level vars, if possible.

canweriotnow commented:
Oh, the joy of NPEs. I should be able to narrow this down pretty easily after the kidlet goes to bed.

canweriotnow commented:
@yurriq I don't think the top-level vars are the issue, although there are some code smells here... I think I see the problem but I'll have to get to my laptop rather than my phone to fix.

borkdude commented:
I suggest writing `(atom {:name “”})` instead of `{:name (atom “”)}` in this source file:
https://github.com/exercism/xclojure/blob/master/exercises/robot-name/src/example.clj
and changing the rest accordingly. More idiomatic Clojure and probably solves this error too.

haus commented on exercises/robot-name/test/robot_name_test.clj:
```diff
@@ -23,4 +23,7 @@
-          "new name is different from old name"))))
+          "new name is different from old name")
+      (is (not= (robot-name/robot-name robbie)
+                (-> robbie robot-name/reset-name robot-name/robot-name))
```
The NPE in travis comes from the threading here:
Passing robbie into `reset-name` calls `swap!` with a function that simply returns the empty string. `swap!` itself returns the value that was swapped in, not the atom. The return of `reset-name` ("") is passed into `robot-name` which expects a map with an atom as the value for a :name key. 
In the let block of robot-name, whatever comes back from `(:name robot)` is derefed using `@`, which will be derefing null, as there isn't a name key in the empty string.

yurrriq commented:
I didn't get to this. Sounds like others have solid ideas.

canweriotnow commented:
Aah, apologies @kytrinyx @nwallace I didn't get a fix in in time to merge this before the deadline. 

FWIW, @borkdude had the right idea; the primary code smell I mentioned earlier was the structure of the state atom in `example.clj`, and I got halfway through a fix and Halloween happened. 

Whoever is in charge of Hacktoberfest, @nwallace should totally get a shirt, he wrote a good test that exposed shoddiness in legacy code. There's no better contribution then that, even if it wasn't mergeable by deadline.

canweriotnow commented:
I will try to rewrite the example based on the state fail (OMG Clojure!) and the tests accordingly in the next day or so... will tag this in the new PR.

nwallace commented:
I took another swing with a bigger refactoring of the tests, and fixed the test breakage.

While I was at it, I also refactored the example implementation ((link here)[https://github.com/exercism/xclojure/pull/151])

P.S. I appreciate the concern for my t-shirt, but the pull requests don't actually have to be accepted -- just submitted ;)

canweriotnow commented:
Looking pretty good to me! If no one else gets to it, I'll do a closer
review and merge in the AM.

On Wed, Nov 2, 2016, 22:16 Nathan Wallace <notifications@github.com> wrote:

> I took another swing with a bigger refactoring of the tests, and fixed the
> test breakage.
>
> While I was at it, I also refactored the example implementation ((link
> here)[#151 <https://github.com/exercism/xclojure/pull/151>])
>
> P.S. I appreciate the concern for my t-shirt, but the pull requests don't
> actually have to be accepted -- just submitted ;)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/xclojure/pull/149#issuecomment-258051316>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXqbRfnWHSzxjOgKcJjgnSw-0uqElEdks5q6UP_gaJpZM4KgtT4>
> .
>
-- 
Jason Lewis
805.5.LAMBDA // decomplecting.org

My GPG key can be found via: gpg --keyserver pgp.mit.edu --recv-keys
0x4514ddcc3cc725de
Key fingerprint: F884 6075 3FC3 E993 6862 F589 4514 DDCC 3CC7 25DE

haus commented:
This looks good to me fwiw.

haus commented:
Thanks for this @nwallace !
